pay/contracts/dropper/HectorDropper.sol

```diff
@@ -0,0 +1,303 @@
+        }
+
+        return drops;
+    }
+
+    function canPerformAirdrop(
+        address from,
+        uint256 index
+    ) external view returns (bool) {
+        if (index >= numberOfAirdrops[from]) return false;
+
+        return airdrops[from][index].releaseTime <= block.timestamp;
+    }
+
+    /* ======== INTERNAL FUNCTIONS ======== */
+
+    function _createAirdrop(
+        address[] calldata tos,
+        uint256 amountPerRecipient,
+        uint48 releaseTime
+    ) internal returns (uint256 index) {
+        uint256 length = tos.length;
+        if (length == 0) revert INVALID_LENGTH();
+        if (amountPerRecipient == 0) revert INVALID_AMOUNT();
+        if (releaseTime <= block.timestamp) revert INVALID_TIME();
+        if (msg.value < fee) revert INSUFFICIENT_FEE();
+
+        for (uint256 i = 0; i < length; i++) {
+            if (tos[i] == address(0)) revert INVALID_ADDRESS();
+        }
+
+        // get the airdrop index
+        index = numberOfAirdrops[msg.sender];
+        numberOfAirdrops[msg.sender] += 1;
+
+        // create a new airdrop
+        Airdrop storage airdrop = airdrops[msg.sender][index];
+        airdrop.from = msg.sender;
+        airdrop.tos = tos;
+        airdrop.amountPerRecipient = amountPerRecipient;
+        airdrop.releaseTime = releaseTime;
+        airdrop.status = AirdropStatus.InProgress;
+        airdrop.fee = fee;
+
+        // fund assets
+        IERC20(token).safeTransferFrom(
+            msg.sender,
+            address(this),
+            amountPerRecipient * length
+        );
+
+        // take fee
+        (bool success, ) = treasury.call{value: fee}('');
+        if (!success) revert FEE_TRANSFER_FAILED();
+    }
+
+    function _cancelAirdrop(uint256 index) internal {
+        if (index >= numberOfAirdrops[msg.sender]) revert INVALID_INDEX();
+
+        Airdrop storage airdrop = airdrops[msg.sender][index];
+        if (airdrop.status != AirdropStatus.InProgress)
+            revert INACTIVE_AIRDROP();
+
+        // update status
+        airdrop.status = AirdropStatus.Cancelled;
+
+        // refund assets
+        IERC20(token).safeTransfer(
+            msg.sender,
+            airdrop.amountPerRecipient * airdrop.tos.length
+        );
+
+        // refund fee
+        (bool success, ) = (msg.sender).call{value: airdrop.fee}('');
+        if (!success) revert FEE_TRANSFER_FAILED();
+    }
+
+    function _releaseAirdrop(address from, uint256 index) internal {
+        if (index >= numberOfAirdrops[from]) revert INVALID_INDEX();
+
+        Airdrop storage airdrop = airdrops[from][index];
+        if (airdrop.status != AirdropStatus.InProgress)
+            revert INACTIVE_AIRDROP();
+        if (airdrop.releaseTime > block.timestamp) revert NOT_RELEASABLE();
+
+        // update status
+        airdrop.status = AirdropStatus.Completed;
+
+        // airdrop assets
+        for (uint256 i = 0; i < airdrop.tos.length; i++) {
+            IERC20(token).safeTransfer(
```

Comment:
@bejaxer need to add the validation to check if the balance is sufficient to airdrop for all wallets

Reply:
Ideally, if there weren't sufficient fund, we'd just reject the operation right from the beginning.

Reply:
> @bejaxer need to add the validation to check if the balance is sufficient to airdrop for all wallets

on createAirdrop function, it does take (amountPerRecipient * tos.length) so the balance is enough for all wallets